lib/money/money.rb

```diff
@@ -1,32 +1,51 @@
-require 'bigdecimal'
-require 'bigdecimal/util'
-
 class Money
   include Comparable
 
-  attr_reader :value, :cents
+  @@mutex = Mutex.new
+  @@zero_money = nil
+
+  DECIMAL_ZERO = BigDecimal.new(0).freeze
+  NUMERIC_REGEX = /\A-?\d*\.?\d*\z/.freeze
+
+  attr_reader :value, :cents, :currency
 
   class << self
-    def new(value = 0, _currency = nil)
-      if value == 0
-        @empty ||= super(0)
-      else
-        super(value, _currency)
+    attr_accessor :parser, :default_currency
+
+    def new(value = nil, currency = default_currency)
+      if value.nil?
+        value = 0
+        deprecate("Support for Money.new(nil) will be removed from the next major revision. Please use Money.new(0) or Money.zero instead.\n")
       end
+      return super(value, currency) unless value == 0
+      @@zero_money || @@mutex.synchronize { @@zero_money = super(0) }
     end
+    alias_method :from_amount, :new
 
     def empty
       new(0)
     end
+    alias_method :zero, :empty
 
-    alias_method :from_amount, :new
+    def parse(input, _currency = default_currency)
+      parser.parse(input)
+    end
+
+    def from_cents(cents, currency = default_currency)
```

Comment:
Do we want to deprecate this and promote something like `from_subunits` instead? That one would be aware of currency instead of dumb division by 100.